File: tools/qr-check.mjs
// Dumps QR matrices for tools/qr-check.py to compare against a reference
// implementation.
//
//   node tools/qr-check.mjs > /tmp/qr.json
//   python3 -m venv /tmp/qrv && /tmp/qrv/bin/pip install qrcode
//   /tmp/qrv/bin/python tools/qr-check.py /tmp/qr.json
//
// The encoder in src/qr.js derives almost everything -- codeword counts, block
// sizes, alignment positions -- from the module layout rather than from tables,
// which is only trustworthy if something independent agrees. This is how the
// error correction table was checked, and how the two bugs that survived
// first-draft testing (the format bits are written most significant first, and
// the module at (n-8, 8) is not a format bit) were found.

import { encode, capacity, LEVELS, MAX_VERSION } from '../src/qr.js';

const ALNUM = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
let seed = 20260908;
const rnd = (n) => {
  seed = (seed * 1103515245 + 12345) & 0x7fffffff;
  return (seed >>> 8) % n;
};

const cases = [];

for (let version = 1; version <= MAX_VERSION; version++) {
  for (const [level, index] of Object.entries(LEVELS)) {
    const room = capacity(version, index).data * 8;
    const countBits = version <= 9 ? 9 : 11;
    // Fill the version nearly to the top: a block structure that is wrong
    // shows up in the last block, not the first.
    const chars = Math.floor((room - 4 - countBits) / 11) * 2;

    let text = '';
    for (let i = 0; i < Math.max(1, chars - rnd(3)); i++) text += ALNUM[rnd(36)];

    for (const mask of [0, 3, 7]) {
      const symbol = encode(text, { level, version, mask });
      cases.push({
        text, version, level, mask,
        size: symbol.size,
        matrix: symbol.matrix.map((row) => row.join('')),
      });
    }
  }
}

process.stdout.write(JSON.stringify(cases));
console.error(`${cases.length} symbols across ${MAX_VERSION} versions`);

File: tools/qr-check.py
#!/usr/bin/env python3
"""Compare tools/qr-check.mjs output against the `qrcode` package.

    python3 -m venv /tmp/qrv && /tmp/qrv/bin/pip install qrcode
    node tools/qr-check.mjs > /tmp/qr.json
    /tmp/qrv/bin/python tools/qr-check.py /tmp/qr.json

Exits non-zero if any symbol differs by a single module. Nothing in squrl
depends on this at runtime or in its tests -- test/qr.mjs carries fixtures that
were verified here, so the check is reproducible without the dependency.
"""

import json
import sys

import qrcode
from qrcode.constants import (
    ERROR_CORRECT_L, ERROR_CORRECT_M, ERROR_CORRECT_Q, ERROR_CORRECT_H,
)

LEVELS = {
    "L": ERROR_CORRECT_L, "M": ERROR_CORRECT_M,
    "Q": ERROR_CORRECT_Q, "H": ERROR_CORRECT_H,
}


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else "/tmp/qr.json"
    cases = json.load(open(path))

    matched = 0
    failures = []

    for case in cases:
        qr = qrcode.QRCode(
            version=case["version"],
            error_correction=LEVELS[case["level"]],
            border=0,
            mask_pattern=case["mask"],
        )
        qr.add_data(case["text"])
        qr.make(fit=False)

        reference = ["".join("1" if cell else "0" for cell in row)
                     for row in qr.get_matrix()]

        if reference == case["matrix"]:
            matched += 1
        else:
            differing = sum(
                1 for a, b in zip("".join(reference), "".join(case["matrix"])) if a != b
            )
            failures.append((case["version"], case["level"], case["mask"], differing))

    print(f"matched {matched} of {len(cases)}")
    for version, level, mask, differing in failures[:20]:
        print(f"  version {version}{level} mask {mask}: {differing} modules differ")

    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())
